### DISCUSSION TITLE
Feature request: a way to regenerate redirects

### DISCUSSION THREAD

**Discussion by jmikrut** [vercel/next.js#38160] [Ideas]
Hi all,

Love NextJS. Love the new on-demand revalidation feature. That's huge.

We have some large projects in production right now that are hooked up to [Payload CMS](https://payloadcms.com), where within, we allow site admins to manage their own redirects. 

The `redirects` functionality within NextJS is performant and beautiful, and works perfectly when combined with Payload.

But.... we need a way to trigger a rebuild of redirects on-the-fly when a redirect is added or changed. Some of our clients are quite active in trying to keep a handle on their 404s and similar, leveraging redirects from a headless CMS, but without a way for them to publish redirects, they need to reach out to us and have us rebuild the entire site whenever they add / change a redirect. 

It would be great if there was a feature that would allow us to programmatically rebuild redirects. Maybe the same could be said for rewrites as well, but our immediate need is only regarding redirects.

That's all - thank you!

**Comment by tilman** [comment]
Hey, we are having the same problem but with strapi cms. Have you found a solution for that? 
I was thinking about automatically triggering a rebuild of the whole site via the CMS and the `/v13/deployments` Vercel REST API.

**Comment by maxjf1** [comment]
I'm having this same issue. We are using an headless CMS that can manage content and redirects automatically. Currently, we are generating the redirects got on the CMS via API on build time. The problem is, each time a new redirect is needed, we need to build the project again.

We had an similar issue with [Incremental Static regeneration](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration), were we needed that whenever an update was made on the CMS, the cache immediately invalidate. With the 12.2 on-demand revalidation feature, we are able to ping back the next server whenever the CMS updates something, allowing to have real-time updated content

The 2 approaches that would be great for this feature:
- Having an *revalidate* argument on the redirects, where after X seconds it regenerates the redirects JSON config without build
- Having an on-demand revalidation feature, where i can call an API route to regenerate the redirects JSON

**Comment by KurtRogers** [comment]
I wanted to share this approach I've taken for a recent project where I used middleware to compare the current path against `/api/redirects?filters[from]=${current_path}` to see if a dynamic redirect existed before sending users to a 404 page. And it works really well. 

This is the Strapi redirect plugin used:
https://market.strapi.io/plugins/strapi-plugin-redirects

`NEXT_FRONTEND_SITE_URL` env var is the `origin` (https://some-website.com)

```tsx 
import { NextResponse, NextRequest } from 'next/server';
import { getFrontendUrl } from '@/app/utils/api-helpers';
import { fetchAPI } from '@/app/utils/fetch-api';

type RedirectEntry = {
  from: string;
  to: string;
  type: StatusType;
};

type StatusType = 'permanent' | 'temporary';

export async function middleware(request: NextRequest) {
  // Get the path for the incoming request
  const pathname = request.nextUrl;

  const { href, pathname: _path } = pathname;

  // check that the page exists and 200's
  const { status: pageStatusResult } = await fetch(
    (process.env.NEXT_FRONTEND_SITE_URL ?? 'http://localhost:3000') + _path
  );

  // if page exists, stop everything here
  if (pageStatusResult === 200) return;

  /**
   *
   * @description
   * if the page 404's, run redirect functionality
   *
   **/
  const redirects = await fetch(
    getFrontendUrl(`/api/redirects?filters[from]=${_path}`)
  );

  const redirectsData = await redirects.json();
  const { data: _redirectData } = redirectsData;

  // kill it here - will take the user to the 404 page
  if (!_redirectData.length) return;

  // we will always get one result back based on the above filter param
  const { to, from, type } = _redirectData[0].attributes as RedirectEntry;

  // run the redirect with the correct status code as defined in the redirect interface in Strapi
  return NextResponse.redirect(
    new URL(to, getFrontendUrl()),
    statusCodeGenerator(type)
  );

  function statusCodeGenerator(type: StatusType) {
    const statusCodes = {
      permanent: 301,
      temporary: 307,
    };

    return statusCodes[type] || statusCodes['permanent'];
  }
}

/**
 *
 * @description
 * built in regex checker for checking against paths
 *
 * Only returns paths which aren't covered by the regex
 * This should always return the root page path we're after (request.nextUrl.pathname)
 *
 */
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     * - all svg files (coming through from the social links in the footer)
     */
    {
      source:
        '/((?!api|robots.txt|_next/static|_next/image|favicon.ico|.*.svg).*)',
      missing: [
        { type: 'header', key: 'next-router-prefetch' },
        { type: 'header', key: 'purpose', value: 'prefetch' },
      ],
    },
  ],
};

```

This is in a production build and is working flawlessly. **Users can update their redirects and they do not need to rebuild the Next app.**

**Reply by rodrigo-arias** [reply]
Thanks for sharing. I think this has a cost since it makes an extra request on every page to check if it is a 404.

**Comment by philwolstenholme** [comment]
We solved this at work using Vercel Edge Config to store the redirects, middleware to read the Edge Config and issue a redirect if appropriate, and a Contentful webhook to update Edge Config when redirects were created/edited/deleted in the CMS. Using Edge Config avoids the cost of a HTTP request to look up the redirects.

Not much help if you're not a Vercel customer, I know :) But it worked really nicely for us.

**Reply by manuel-schoebel** [reply]
@philwolstenholme - I am facing the same challenge and tried out edge config. But it seems that this would only work with very few redirects. I need to maintain "a couple of thousands" redirects. When I want to store this in the edge config it is simply too large of a document.

Also in the edge config you cannot store e.g. the "from" path like  /some/page as a key so you need to hash this. But then in the middleware you cannot use crypto so you need to do some simple custom hash fn. And all of this feels totally wrong.

I would be very interested in your approach. Maybe you can add a bit more details. Also how many redirects do you maintain? Are you on pro or enterprise plan? How do you store the redirect info on the edge config? 😊
